sycl/doc/SpecializationConstants.md

```diff
@@ -0,0 +1,307 @@
+
+```cpp
+struct A {
+  int x;
+  float y;
+};
+
+struct POD {
+  A a[2];
+  int b;
+};
+```
+
+and the user says
+
+```cpp
+  POD gold{
+    {
+      { goldi, goldf },
+      { goldi + 1, goldf + 1 },
+    },
+    goldi
+  };
+
+  cl::sycl::ONEAPI::experimental::spec_constant<POD, MyConst> sc =  program4.set_spec_constant<MyConst>(gold);
+```
+
+#### Compiler
+
+##### The SpecConstant pass changes
+
+ - The SpecConstants pass in the post-link will have the following IR as input (`sret` conversion is omitted for clarity):
+
+```cpp
+  %spec_const = call %struct.POD __sycl_getCompositeSpecConstantValue<POD type mangling> ("MyConst_mangled")
+```
+
+where `__sycl_getCompositeSpecConstantValue` is a new "intrinsic"
+ (in addition to `__sycl_getSpecConstantValue`) recognized by SpecConstants pass,
+ which creates a value of a composite (of non-primitive type) specialization constant.
+ It does not need a default value, because its default value consists of default
+ valued of its leaf specialization constants (see below).
+
+ - after spec constant enumeration (symbolic -\> int ID translation), the SpecConstants pass
+ will handle the `__sycl_getCompositeSpecConstantValue`. Given the knowledge of the composite
+ specialization constant's type (`%struct.POD`), the pass will traverse its leaf
+ fields and generate 5 "primitive" spec constants using already existing SPIR-V intrinsic:
+
+```cpp
+%gold_POD_a0x = call i32 __spirv_SpecConstant(i32 10, i32 0)
+%gold_POD_a0y = call float __spirv_SpecConstant(i32 11, float 0)
+%gold_POD_a1x = call i32 __spirv_SpecConstant(i32 12, i32 0)
+%gold_POD_a1y = call float __spirv_SpecConstant(i32 13, float 0)
+%gold_POD_b = call i32 __spirv_SpecConstant(i32 14, i32 0)
+```
+
+And 1 "composite"
+
+```cpp
+  %gold_POD = call %struct.POD __spirvCompositeSpecConstant<POD type mangling>(i32 10, i32 11, i32 12, i32 13, i32 14, i32 15)
+```
+
+where `__spirvCompositeSpecConstant<POD type mangling>` is a new SPIR-V intrinsic which
+ represents creation of a composite specialization constant. Its arguments are spec
+ constant IDs corresponding to the leaf fields of the POD type of the constant.
+Spec ID for the composite spec constant is not needed, as runtime will never use it - it will use IDs of the leaves instead.
+ Yet SPIR-V does not allow IDs for composite spec constants.
+
+##### The post-link tool changes
+
+For composite specialization constants the post link tool will additionally
+generate linearized list of \<leaf spec ID,type,offset,size\> tuples (descriptors),
+where each tuple describes a leaf field, and store it together with the
+existing meta-information associated with the specialization constants and
+passed to the runtime. Also, for a composite specialization constant the will be
+no ID map entry within the meta information, and the composite constant will
+referenced by its symbolic ID. For example:
+
+```cpp
+MyConst_mangled [10,int,0,4],[10,float,4,4],[10,int,8,4],[10,float,12,4],[10,int,16,4]
```

Comment:
```suggestion
MyConst_mangled [10,int,0,4],[11,float,4,4],[12,int,8,4],[13,float,12,4],[14,int,16,4]
```
If I understand it correctly, this is data structure(list) describes the byte array given as an argument to `program4.set_spec_constant<MyConst>(gold)`. If so, I think the type field is redundant here. Having size and offset should be enough to read the value from the byte array.